lib/kennel/parts_serializer.rb

```diff
@@ -35,13 +35,20 @@ def write_changed(parts)
     end
 
     def existing_files_and_folders
-      if filter.tracking_id_filter
-        filter.tracking_id_filter.map { |tracking_id| path_for_tracking_id(tracking_id) }
-      elsif filter.project_filter
-        filter.project_filter.flat_map { |project| Dir["generated/#{project}/*"] }
-      else
-        Dir["generated/**/*"] # also includes folders so we clean up empty directories
+      paths = Dir["generated/**/*"]
+
+      if filter.filtering?
+        segment = Kennel::Models::Record::ALLOWED_KENNEL_ID_SEGMENT
+        paths.select! do |path|
+          if (m = path.match(/^generated\/(#{segment})\/(#{segment})\.json$/o))
+            filter.matches_tracking_id?("#{m[1]}:#{m[2]}")
+          elsif (m = path.match(/^generated\/(#{segment})(?:\/|$)/o)) && File.directory?(path)
+            filter.matches_project_id?(m[1])
+          end
+        end
```

Comment:
same ?
```suggestion
        paths.select! do |path|
          tracking_id = path.split("/")[1..2].to_a.join(":")
          filter.matches_tracking_id?(tracking_id)
        end
```

Reply:
What you wrote doesn't causes any existing tests to fail, but I'm still reluctant, because it means that `tracking_id` ends up being things which aren't (i.e. can't possibly be) tracking IDs. e.g. `foo` or `a space!`.

It also makes me think that we're missing a test somewhere, but I'm not sure what yet. 🤔

Reply:
Accepting this until I can think of a more concrete thing to add. :-)